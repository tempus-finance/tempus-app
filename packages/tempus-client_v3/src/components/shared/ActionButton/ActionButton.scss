.tc__button {
  padding: 8px;
  border-radius: 4px;
  box-shadow: var(--downWeakShadow);
  box-sizing: border-box;
  mix-blend-mode: normal;
  cursor: pointer;
}
