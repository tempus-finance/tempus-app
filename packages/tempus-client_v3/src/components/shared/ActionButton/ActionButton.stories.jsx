import ActionButton from './ActionButton';

export default {
  title: 'ActionButton',
  component: ActionButton,
  argTypes: {
    labels: {
      control: {
        type: 'object',
      },
    },
    variant: {
      control: {
        type: 'select',
        options: ['primary', 'secondary', 'tertiary'],
      },
    },
    size: {
      control: {
        type: 'select',
        options: ['small', 'large'],
      },
    },
    state: {
      control: {
        type: 'select',
        options: ['default', 'disabled', 'loading', 'success'],
      },
    },
    fullWidth: {
      control: {
        type: 'boolean',
      },
    },
  },
};

const style = {
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  width: '100%',
  height: '150px',
};

const Template = args => (
  <div style={style}>
    <ActionButton {...args} />
  </div>
);

export const Primary = Template.bind({});
Primary.args = {
  labels: {
    default: 'Action Button',
    loading: 'Loading...',
    success: 'Success',
  },
  variant: 'primary',
  size: 'small',
  state: 'default',
  fullWidth: false,
};
