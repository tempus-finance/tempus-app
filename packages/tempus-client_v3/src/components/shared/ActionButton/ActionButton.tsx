import { FC, memo } from 'react';
import Button from '../Button';
import Icon from '../Icon';
import Loading, { LoadingColor } from '../Loading';
import Typography, { TypographyColor, TypographyWeight } from '../Typography';

import './ActionButton.scss';

type ButtonSize = 'small' | 'large';
type ButtonVariant = 'primary' | 'secondary' | 'tertiary';
type ButtonState = 'default' | 'disabled' | 'loading' | 'success';
type ButtonLabels = {
  default: string;
  loading: string;
  success: string;
};

interface ButtonProps {
  labels: ButtonLabels;
  onClick: () => void;
  variant?: ButtonVariant;
  state?: ButtonState;
  size?: ButtonSize;
  fullWidth?: boolean;
}

const ActionButton: FC<ButtonProps> = props => {
  const { labels, variant = 'primary', size = 'small', state = 'default', fullWidth = false, onClick } = props;

  // Compute CSS class names
  let borderClass = `tc__actionButton-border-${variant}-${size}`;
  let backgroundClass = `tc__actionButton-background-${variant}-${size}`;
  let hoverClass = `tc__actionButton-hover-${variant}-${size}`;
  let shadowClass = `tc__actionButton-shadow-${state}`;
  let cursorClass = `tc__actionButton-cursor-${state}`;
  let heightClass = `tc__actionButton-size-${size}`;
  let widthClass = fullWidth ? 'tc__actionButton-width-full' : '';
  switch (state) {
    case 'disabled':
      borderClass += '-disabled';
      backgroundClass += '-disabled';
      hoverClass += '-disabled';
      break;
    case 'loading':
      borderClass += '-loading';
      backgroundClass += '-loading';
      hoverClass += '-loading';
      break;
    case 'success':
      borderClass += '-success';
      backgroundClass += '-success';
      hoverClass += '-success';
      break;
  }

  // Text color
  let textColor: TypographyColor = 'text-primary';
  switch (variant) {
    case 'primary':
      textColor = 'text-primary-inverted';
      break;
  }
  switch (state) {
    case 'success':
      textColor = 'text-success';
  }
  if (variant === 'secondary' && state === 'disabled') {
    textColor = 'text-disabled';
  }
  if (variant === 'secondary' && state === 'loading') {
    textColor = 'text-disabled-secondary';
  }
  if (variant === 'tertiary' && (state === 'disabled' || state === 'loading')) {
    textColor = 'text-disabled-secondary';
  }

  // Text opacity
  let textOpacity: number = 1;
  if (variant === 'primary' && state === 'disabled') {
    textOpacity = 0.5;
  }

  // Text weight
  let textWeight: TypographyWeight = 'bold';
  switch (state) {
    case 'disabled':
      textWeight = 'regular';
      break;
    case 'loading':
      textWeight = 'medium';
      break;
  }

  // Loader color
  let loaderColor: LoadingColor = 'default';
  if ((variant === 'secondary' || variant === 'tertiary') && state === 'loading') {
    loaderColor = 'secondary';
  }

  return (
    <Button
      className={`tc__actionButton ${borderClass} ${backgroundClass} ${hoverClass} ${shadowClass} ${cursorClass} ${heightClass} ${widthClass}`}
      disabled={state === 'disabled'}
      onClick={onClick}
      style={{
        width: '20000px', // fullWidth ? '100%' : '',
      }}
    >
      {state === 'loading' && (
        <div className={`tc__actionButton-status-${size}-content`}>
          <Loading size={size === 'small' ? 12 : 20} color={loaderColor} />
        </div>
      )}
      {state === 'success' && (
        <div className={`tc__actionButton-status-${size}-content`}>
          <Icon type="checkmark-solid" size={size === 'small' ? 12 : 20} />
        </div>
      )}
      <div className="tc__actionButton-status-label">
        <Typography
          variant={size === 'large' ? 'body-primary' : 'body-secondary'}
          weight={textWeight}
          color={textColor}
          opacity={textOpacity}
        >
          {(state === 'default' || state === 'disabled') && labels.default}
          {state === 'loading' && labels.loading}
          {state === 'success' && labels.success}
        </Typography>
      </div>
    </Button>
  );
};

export default memo(ActionButton);
