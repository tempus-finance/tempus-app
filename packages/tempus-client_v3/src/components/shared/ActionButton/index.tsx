import ActionButton, { ActionButtonLabels, ActionButtonState, ActionButtonVariant } from './ActionButton';

export default ActionButton;
export type { ActionButtonLabels, ActionButtonState, ActionButtonVariant };
