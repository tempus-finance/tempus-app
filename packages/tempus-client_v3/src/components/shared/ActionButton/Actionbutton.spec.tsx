import { fireEvent, render } from '@testing-library/react';
import ActionButton, { ButtonProps } from './ActionButton';

const mockOnClick = jest.fn();

const defaultProps: ButtonProps = {
  labels: { default: 'Default', loading: 'Loading', success: 'Success' },
  onClick: mockOnClick,
};

const subject = (props: ButtonProps) => render(<ActionButton {...props} />);

describe('ActionButton', () => {
  it('renders a button with a `default` label', () => {
    const { queryByText, getByRole } = subject(defaultProps);

    const actualButton = getByRole('button');
    const actualText = queryByText('Default');

    expect(actualText).not.toBeNull();

    expect(actualButton).toMatchSnapshot();
  });

  it('renders a button with a `loading` label', () => {
    const { queryByText, getByRole } = subject({ ...defaultProps, state: 'loading' });

    const actualButton = getByRole('button');
    const actualText = queryByText('Loading');

    expect(actualText).not.toBeNull();

    expect(actualButton).toMatchSnapshot();
  });

  it('renders a button with a `success` label', () => {
    const { queryByText, getByRole } = subject({ ...defaultProps, state: 'success' });

    const actualButton = getByRole('button');
    const actualText = queryByText('Success');

    expect(actualText).not.toBeNull();

    expect(actualButton).toMatchSnapshot();
  });

  it('renders a button with a `disabled` state', () => {
    const { queryByText, getByRole } = subject({ ...defaultProps, state: 'disabled' });

    const actualButton = getByRole('button');
    const actualText = queryByText('Default');

    expect(actualText).not.toBeNull();

    expect(actualButton).toMatchSnapshot();
  });

  it('calls `onClick` when the button is clicked', () => {
    const { getByRole } = subject(defaultProps);

    const button = getByRole('button');
    fireEvent.click(button);

    expect(mockOnClick).toHaveBeenCalledTimes(1);
  });
});
