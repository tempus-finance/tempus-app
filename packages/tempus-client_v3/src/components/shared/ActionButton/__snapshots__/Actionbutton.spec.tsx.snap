// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionButton renders a button with a \`default\` label 1`] = `
<button
  class="tc__btn tc__actionButton tc__actionButton-border-primary-small tc__actionButton-background-primary-small tc__actionButton-hover-primary-small tc__actionButton-shadow-default tc__actionButton-cursor-default tc__actionButton-size-small "
  data-selected="false"
>
  <div
    class="tc__actionButton-status-label"
  >
    <div
      style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(255, 255, 255); font-weight: 700; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      Default
    </div>
  </div>
</button>
`;

exports[`ActionButton renders a button with a \`disabled\` state 1`] = `
<button
  class="tc__btn tc__actionButton tc__actionButton-border-primary-small-disabled tc__actionButton-background-primary-small-disabled tc__actionButton-hover-primary-small-disabled tc__actionButton-shadow-disabled tc__actionButton-cursor-disabled tc__actionButton-size-small "
  data-selected="false"
  disabled=""
>
  <div
    class="tc__actionButton-status-label"
  >
    <div
      style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(255, 255, 255); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 0.5;"
    >
      Default
    </div>
  </div>
</button>
`;

exports[`ActionButton renders a button with a \`loading\` label 1`] = `
<button
  class="tc__btn tc__actionButton tc__actionButton-border-primary-small-loading tc__actionButton-background-primary-small-loading tc__actionButton-hover-primary-small-loading tc__actionButton-shadow-loading tc__actionButton-cursor-loading tc__actionButton-size-small "
  data-selected="false"
>
  <div
    class="tc__actionButton-status-small-content"
  >
    <svg
      height="12"
      viewBox="0 0 100 100"
      width="12"
    >
      <circle
        class="tc__loading tc__loading__color-bg-default"
        cx="50"
        cy="50"
        r="40"
      />
      <circle
        class="tc__loading tc__loading__color-animate-default tc__loading__animate"
        cx="50"
        cy="50"
        r="40"
      />
    </svg>
  </div>
  <div
    class="tc__actionButton-status-label"
  >
    <div
      style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(255, 255, 255); font-weight: 500; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      Loading
    </div>
  </div>
</button>
`;

exports[`ActionButton renders a button with a \`success\` label 1`] = `
<button
  class="tc__btn tc__actionButton tc__actionButton-border-primary-small-success tc__actionButton-background-primary-small-success tc__actionButton-hover-primary-small-success tc__actionButton-shadow-success tc__actionButton-cursor-success tc__actionButton-size-small "
  data-selected="false"
>
  <div
    class="tc__actionButton-status-small-content"
  >
    <svg
      class="tc__icon tc__icon-checkmark-solid"
      fill="none"
      height="12"
      viewBox="0 0 16 16"
      width="12"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="8"
        cy="8"
        fill="#4BB543"
        r="8"
      />
      <path
        d="M4 8.05015L7 11L12 6L11 5L7 9L5 7L4 8.05015Z"
        fill="white"
      />
    </svg>
  </div>
  <div
    class="tc__actionButton-status-label"
  >
    <div
      style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(75, 181, 67); font-weight: 700; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      Success
    </div>
  </div>
</button>
`;

exports[`ActionButton renders a secondary button with a \`default\` label 1`] = `
<button
  class="tc__btn tc__actionButton tc__actionButton-border-secondary-small tc__actionButton-background-secondary-small tc__actionButton-hover-secondary-small tc__actionButton-shadow-default tc__actionButton-cursor-default tc__actionButton-size-small tc__actionButton-width-full"
  data-selected="false"
>
  <div
    class="tc__actionButton-status-label"
  >
    <div
      style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(34, 34, 34); font-weight: 700; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      Default
    </div>
  </div>
</button>
`;

exports[`ActionButton renders a secondary button with a \`disabled\` state 1`] = `
<button
  class="tc__btn tc__actionButton tc__actionButton-border-secondary-small-disabled tc__actionButton-background-secondary-small-disabled tc__actionButton-hover-secondary-small-disabled tc__actionButton-shadow-disabled tc__actionButton-cursor-disabled tc__actionButton-size-small "
  data-selected="false"
  disabled=""
>
  <div
    class="tc__actionButton-status-label"
  >
    <div
      style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(122, 122, 122); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      Default
    </div>
  </div>
</button>
`;

exports[`ActionButton renders a secondary button with a \`loading\` label 1`] = `
<button
  class="tc__btn tc__actionButton tc__actionButton-border-secondary-small-loading tc__actionButton-background-secondary-small-loading tc__actionButton-hover-secondary-small-loading tc__actionButton-shadow-loading tc__actionButton-cursor-loading tc__actionButton-size-small "
  data-selected="false"
>
  <div
    class="tc__actionButton-status-small-content"
  >
    <svg
      height="12"
      viewBox="0 0 100 100"
      width="12"
    >
      <circle
        class="tc__loading tc__loading__color-bg-secondary"
        cx="50"
        cy="50"
        r="40"
      />
      <circle
        class="tc__loading tc__loading__color-animate-secondary tc__loading__animate"
        cx="50"
        cy="50"
        r="40"
      />
    </svg>
  </div>
  <div
    class="tc__actionButton-status-label"
  >
    <div
      style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(191, 192, 192); font-weight: 500; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      Loading
    </div>
  </div>
</button>
`;

exports[`ActionButton renders a secondary button with a \`success\` label 1`] = `
<button
  class="tc__btn tc__actionButton tc__actionButton-border-secondary-large-success tc__actionButton-background-secondary-large-success tc__actionButton-hover-secondary-large-success tc__actionButton-shadow-success tc__actionButton-cursor-success tc__actionButton-size-large "
  data-selected="false"
>
  <div
    class="tc__actionButton-status-large-content"
  >
    <svg
      class="tc__icon tc__icon-checkmark-solid"
      fill="none"
      height="20"
      viewBox="0 0 16 16"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="8"
        cy="8"
        fill="#4BB543"
        r="8"
      />
      <path
        d="M4 8.05015L7 11L12 6L11 5L7 9L5 7L4 8.05015Z"
        fill="white"
      />
    </svg>
  </div>
  <div
    class="tc__actionButton-status-label"
  >
    <div
      style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(75, 181, 67); font-weight: 700; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      Success
    </div>
  </div>
</button>
`;

exports[`ActionButton renders a tertiary button with a \`disabled\` state 1`] = `
<button
  class="tc__btn tc__actionButton tc__actionButton-border-tertiary-small-disabled tc__actionButton-background-tertiary-small-disabled tc__actionButton-hover-tertiary-small-disabled tc__actionButton-shadow-disabled tc__actionButton-cursor-disabled tc__actionButton-size-small "
  data-selected="false"
  disabled=""
>
  <div
    class="tc__actionButton-status-label"
  >
    <div
      style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(191, 192, 192); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      Default
    </div>
  </div>
</button>
`;

exports[`ActionButton renders a tertiary button with a \`loading\` state 1`] = `
<button
  class="tc__btn tc__actionButton tc__actionButton-border-tertiary-large-loading tc__actionButton-background-tertiary-large-loading tc__actionButton-hover-tertiary-large-loading tc__actionButton-shadow-loading tc__actionButton-cursor-loading tc__actionButton-size-large "
  data-selected="false"
>
  <div
    class="tc__actionButton-status-large-content"
  >
    <svg
      height="20"
      viewBox="0 0 100 100"
      width="20"
    >
      <circle
        class="tc__loading tc__loading__color-bg-secondary"
        cx="50"
        cy="50"
        r="40"
      />
      <circle
        class="tc__loading tc__loading__color-animate-secondary tc__loading__animate"
        cx="50"
        cy="50"
        r="40"
      />
    </svg>
  </div>
  <div
    class="tc__actionButton-status-label"
  >
    <div
      style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(191, 192, 192); font-weight: 500; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      Loading
    </div>
  </div>
</button>
`;
